### PR TITLE
Introduce `ConflictDetection` utility class

### PR DESCRIPTION
**Pull request by eric-picnic**
Suggested commit message:
```
Introduce support for detecting conflicts when renaming methods (#478)

By moving reusable conflict detection code from `JUnitMethodDeclaration` to a
new `ConflictDetection` class.
```

### CODE REVIEW

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ConflictDetection.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ImportTree;
+import com.sun.source.tree.Tree;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+import com.sun.tools.javac.code.Type;
+import java.util.Optional;
+
+/**
+ * A set of helper methods for finding conflicts which would be caused by applying certain fixes.
```
Comment (rickie) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ConflictDetection.java]: Hmm, not a huge fan of the "certain fixes".

I'm suggesting an alternative,feel free to revert or tweak :).

File: error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/util/ConflictDetectionTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            "",
+            "public class B {",
+            "  public static void foo3t() {}",
+            "}")
+        .doTest();
+  }
+
+  /**
+   * A {@link BugChecker} that flags method rename blockers found by {@link
+   * ConflictDetection#findMethodRenameBlocker(Symbol.MethodSymbol, String, VisitorState)}.
```
Comment (rickie) [inline error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/util/ConflictDetectionTest.java]:
```suggestion
   * ConflictDetection#findMethodRenameBlocker(MethodSymbol, String, VisitorState)}.
```

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/JUnitMethodDeclaration.java [hunk 3]
```diff
@@ -8,8 +8,8 @@
 import static com.google.errorprone.matchers.Matchers.hasModifier;
 import static com.google.errorprone.matchers.Matchers.not;
 import static java.util.function.Predicate.not;
+import static tech.picnic.errorprone.bugpatterns.util.ConflictDetection.findMethodRenameBlocker;
```
Comment (rickie) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/JUnitMethodDeclaration.java]: Let's not statically import. That may read a bit better.

**Review by rickie — APPROVED** [review]
Added a commit with some suggestions. 

Suggested commit message;
```
Introduce `ConflictDetection` utility class (#478)
```

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ConflictDetection.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ImportTree;
+import com.sun.source.tree.Tree;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+import com.sun.tools.javac.code.Type;
+import java.util.Optional;
+
+/** A set of helper methods for detecting conflicts that would be caused when applying fixes. */
+public final class ConflictDetection {
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ConflictDetection.java:14]: Fine for now, but I suspect that eventually we'll want a more generic name. Time will tell :)

**Review by Stephan202 — APPROVED** [review]
Rebased and added a commit. Didn't call out all changes.

**Comment by Stephan202** [comment]
I see that one mutant survived; will have a look at that in a bit.

File: error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/util/ConflictDetectionTest.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+   * ConflictDetection#findMethodRenameBlocker(MethodSymbol, String, VisitorState)}.
+   */
+  @BugPattern(summary = "Flags blockers for renaming methods", severity = ERROR)
+  public static final class RenameBlockerFlagger extends BugChecker implements MethodTreeMatcher {
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    public Description matchMethod(MethodTree tree, VisitorState state) {
+      return ConflictDetection.findMethodRenameBlocker(
+              ASTHelpers.getSymbol(tree), tree.getName() + "t", state)
```
[the remark this answers is not in the document]
Comment (rickie) [reply error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/util/ConflictDetectionTest.java:68]: Yes but then indeed the `in` -> `int` case is not tested nicely. Couldn't figure out a nicer way to do that. (Based on the changes not including this change I suspect you went through the same train of thoughts there 😉).

**Review by rickie — COMMENTED** [review]
Changes LGTM!

**Comment by rickie** [comment]
Rebased, will merge once 🍏.

**Pull request merged**